### PR TITLE
ACS-3377 Update rule order.

### PR DESCRIPTION
**Pull request by tpage-alfresco**

### CODE REVIEW

File: remote-api/src/main/java/org/alfresco/rest/api/impl/rules/RuleSetsImpl.java [hunk 1]
```diff
@@ -80,6 +85,43 @@ public RuleSet getRuleSetById(String folderNodeId, String ruleSetId, List<String
...
+            Set<String> suppliedRuleIdSet = new HashSet<>(suppliedRuleIds);
+
+            // Check that the set of rule ids hasn't changed.
+            Set<String> existingRuleIds = new HashSet<>(ruleSetLoader.loadRuleIds(folderNode));
+            if (!suppliedRuleIdSet.equals(existingRuleIds))
+            {
+                throw new InvalidArgumentException("Unexpected set of rule ids - received " + suppliedRuleIds + " but expected " + existingRuleIds);
+            }
+
+            IntStream.range(0, suppliedRuleIds.size()).forEach(index ->
```
[the remark this answers is not in the document]
Comment (krdabrowski) [reply remote-api/src/main/java/org/alfresco/rest/api/impl/rules/RuleSetsImpl.java:111]: Hmm... unit test covering this scenario should be enough.

File: packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/rules/ReorderRules.java [hunk 2]
```diff
@@ -77,4 +78,127 @@ public void getOrderedRuleIds()
...
+        restClient.assertStatusCodeIs(OK);
+        ruleSet.assertThat().field("ruleIds").is(reversedRuleIds);
+    }
+
+    /** Create three rules in the given folder. */
+    private List<RestRuleModel> createRulesInFolder(FolderModel folder, UserModel user)
+    {
+        return IntStream.range(0, 3).mapToObj(index ->
+        {
+            RestRuleModel ruleModel = createRuleModel("ruleName");
```
[the remark this answers is not in the document]
Comment (tpage-alfresco) [reply packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/rules/ReorderRules.java:200]: It's called `createRuleModelWithDefaultValues` which is somehow longer than `createRuleModel("ruleName")` :joy:

File: remote-api/src/main/java/org/alfresco/rest/api/impl/rules/RuleSetsImpl.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (tpage-alfresco) [reply remote-api/src/main/java/org/alfresco/rest/api/impl/rules/RuleSetsImpl.java:111]: You're right. I added the comment `Check there are no duplicate rule ids in the request` above and then forgot to actually add the validation.

File: packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/rules/ReorderRules.java [hunk 3]
```diff
@@ -77,4 +78,127 @@ public void getOrderedRuleIds()
...
+        RestRuleSetModel ruleSetBody = new RestRuleSetModel();
+        ruleSetBody.setId("-default-");
+        ruleSetBody.setRuleIds(reversedRuleIds);
+        RestRuleSetModel ruleSet = restClient.authenticateUser(user).withCoreAPI().usingNode(ruleFolder)
+                                             .include("ruleIds").updateRuleSet(ruleSetBody);
+
+        restClient.assertStatusCodeIs(FORBIDDEN);
+    }
+
+    /** Check that a user cannot reorder the rules in a rule set if they only have read permission. */
```
[the remark this answers is not in the document]
Comment (tpage-alfresco) [reply packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/rules/ReorderRules.java:170]: Fixed - thanks

**Review by krdabrowski — APPROVED** [review]
LGTM

**Pull request merged**